src/core.rs

```diff
@@ -0,0 +1,247 @@
+/// The main hash code used by Starknet.
+///
+/// Contains 251 bits of data and is generated by the pedersen hash function.
+#[derive(Clone, Debug, PartialEq)]
+pub struct StarkHash([u8; 32]);
+
+/// Error returned by [StarkHash::from_hex_str]
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum HexParseError {
+    /// Hex-string contained a non-hex character.
+    InvalidNibble(char),
+    /// Hex-string contained more than 251 bits of data.
+    Overflow,
+}
+
+impl StarkHash {
+    /// Parses a hex string into a [StarkHash].
+    ///
+    /// Supports the following:
+    ///     - optional "0x" prefix
+    ///     - upper and lower case ('A'-'F' and 'a'-'f').
+    ///     - odd nibbles (e.g. 0x123 == 0x0123)
+    ///     - short sequences (less than 64 nibbles)
+    ///     - leading zeros
+    ///
+    /// Returns [HexParseError::InvalidNibble] if the sequence contains a non-hex
+    /// character.
+    ///
+    /// Returns [HexParseError::Overflow] if the sequence exceeds 251 bits of data.
+    pub fn from_hex_str(hex_str: &str) -> Result<Self, HexParseError> {
+        let hex_str = hex_str.strip_prefix("0x").unwrap_or(hex_str);
+        if hex_str.len() > 64 {
+            return Err(HexParseError::Overflow);
+        }
+
+        let mut buf = [0u8; 32];
+
+        // We want the result in big-endian so reverse iterate over each pair of nibbles.
+        let chunks = hex_str.as_bytes().rchunks_exact(2);
+
+        // Handle a possible odd nibble remaining nibble.
+        let odd_nibble = chunks.remainder();
+        if !odd_nibble.is_empty() {
+            let full_bytes = hex_str.len() / 2;
+            buf[31 - full_bytes] = parse_hex_digit(odd_nibble[0])?;
+        }
+
+        for (i, c) in chunks.enumerate() {
+            // Indexing c[0] and c[1] are safe since chunk-size is 2.
+            buf[31 - i] = parse_hex_digit(c[0])? << 4 | parse_hex_digit(c[1])?;
+        }
+
+        // Only 251 bits are allowed to be used.
+        if buf[0] & 0b1111_1000 > 0 {
+            return Err(HexParseError::Overflow);
+        }
+
+        Ok(Self(buf))
+    }
+
+    /// Returns the big-endian representation of this [StarkHash].
+    pub fn to_be_bytes(self) -> [u8; 32] {
+        self.0
+    }
+
+    /// Creates a [StarkHash] from big-endian bytes.
+    ///
+    /// #### Warning: does not check that only 251 bits are set!
+    pub fn from_be_bytes(bytes: [u8; 32]) -> Self {
+        Self(bytes)
```

Comment:
Why doesn't this check the overflow? I think there's a sort of symmetry issue with the from_hex_str.